File: src/playback/__tests__/playbackRequestPolicy.test.js
import assert from 'node:assert/strict';
import { test } from 'node:test';

import { createPlaybackRequestPolicy } from '../playbackRequestPolicy.js';

test('should discover iOS sources without progressive direct stream', () => {
    assert.deepEqual(createPlaybackRequestPolicy('ios'), {
        enableDirectPlay: true,
        enableDirectStream: false,
        allowVideoStreamCopy: true,
        allowAudioStreamCopy: true
    });
});

test('should force non-direct-play iOS sources through encoded HLS', () => {
    assert.deepEqual(createPlaybackRequestPolicy('ios', {
        IsRemote: false,
        SupportsDirectPlay: false
    }), {
        enableDirectPlay: false,
        enableDirectStream: false,
        allowVideoStreamCopy: false,
        allowAudioStreamCopy: false
    });
});

test('should preserve direct play for compatible local iOS sources', () => {
    assert.deepEqual(createPlaybackRequestPolicy('ios', {
        IsRemote: false,
        SupportsDirectPlay: true
    }), {
        enableDirectPlay: true,
        enableDirectStream: false,
        allowVideoStreamCopy: true,
        allowAudioStreamCopy: true
    });
});

test('should preserve Android direct and remux capabilities', () => {
    assert.deepEqual(createPlaybackRequestPolicy('android', { IsRemote: true }), {
        enableDirectPlay: true,
        enableDirectStream: true,
        allowVideoStreamCopy: true,
        allowAudioStreamCopy: true
    });
});

File: apps/android-tv/feature/player/src/test/kotlin/app/homeflix/tv/feature/player/PlaybackRequestPolicyTest.kt
package app.homeflix.tv.feature.player

import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class PlaybackRequestPolicyTest {
    @Test
    fun `should disable direct stream during discovery`() {
        val policy = androidTvPlaybackPolicy()

        assertTrue(policy.enableDirectPlay)
        assertFalse(policy.enableDirectStream)
        assertTrue(policy.enableTranscoding)
        assertTrue(policy.allowVideoStreamCopy)
        assertTrue(policy.allowAudioStreamCopy)
    }

    @Test
    fun `should force hls for remote source`() {
        val policy = androidTvPlaybackPolicy(mediaSource(isRemote = true))

        assertFalse(policy.enableDirectPlay)
        assertFalse(policy.enableDirectStream)
        assertTrue(policy.enableTranscoding)
    }

    @Test
    fun `should retain exact direct play for local source`() {
        val policy = androidTvPlaybackPolicy(mediaSource(isRemote = false))

        assertTrue(policy.enableDirectPlay)
        assertFalse(policy.enableDirectStream)
        assertTrue(policy.enableTranscoding)
    }

    @Test
    fun `should force hls when source locality is unknown`() {
        val policy = androidTvPlaybackPolicy(mediaSource(isRemote = null))

        assertFalse(policy.enableDirectPlay)
        assertFalse(policy.enableDirectStream)
    }

    private fun mediaSource(isRemote: Boolean?): MediaSourceDto =
        MediaSourceDto(
            id = "source-1",
            name = "Source",
            supportsDirectPlay = true,
            supportsDirectStream = true,
            supportsTranscoding = true,
            transcodingUrl = "/Videos/item-1/master.m3u8",
            transcodingSubProtocol = "hls",
            mediaStreams = emptyList(),
            isRemote = isRemote,
        )
}
